go-controller/pkg/ovn/egressfirewall_dns.go

```diff
@@ -0,0 +1,220 @@
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/klog"
+	utilnet "k8s.io/utils/net"
+)
+
+type EgressDNS struct {
+	// Protects pdMap/namespaces operations
+	lock sync.Mutex
+	// holds DNS entries globally
+	dns *util.DNS
+	// this map holds dnsNames to the dnsEntries
+	dnsEntries map[string]*dnsEntry
+	// allows for the creation of addresssets
+	addressSetFactory AddressSetFactory
+
+	// Report change when Add operation is done
+	added          chan string
+	stopChan       chan struct{}
+	controllerStop <-chan struct{}
+}
+
+type dnsEntry struct {
+	//this map holds all the namespaces that a dnsName appears in
+	namespaces map[string]struct{}
+	// the current IP addresses the dnsName resolves to
+	dnsResolves []net.IP
+	// the addressSet that contains the current IPs
+	dnsAddressSet AddressSet
+}
+
+func NewEgressDNS(addressSetFactory AddressSetFactory, controllerStop <-chan struct{}) (*EgressDNS, error) {
+	dnsInfo, err := util.NewDNS("/etc/resolv.conf")
+	if err != nil {
+		utilruntime.HandleError(err)
+		return nil, err
+	}
+
+	egressDNS := &EgressDNS{
+		dns:               dnsInfo,
+		dnsEntries:        make(map[string]*dnsEntry),
+		addressSetFactory: addressSetFactory,
+
+		added:          make(chan string, 1),
+		stopChan:       make(chan struct{}),
+		controllerStop: controllerStop,
+	}
+
+	go egressDNS.Sync()
+
+	return egressDNS, nil
+}
+
+func (e *EgressDNS) Add(namespace, dnsName string) (AddressSet, error) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	if _, exists := e.dnsEntries[dnsName]; !exists {
+		var err error
+		dnsEntry := dnsEntry{
+			namespaces: make(map[string]struct{}),
+		}
+		if e.addressSetFactory == nil {
+			panic("addressSetFactory is null, that cannot happen")
+		}
+		dnsEntry.dnsAddressSet, err = e.addressSetFactory.NewAddressSet(dnsName, nil)
+		if err != nil {
+			return nil, fmt.Errorf("cannot create addressSet for %s: %v", dnsName, err)
+		}
+		e.dnsEntries[dnsName] = &dnsEntry
+		e.signalAdded(dnsName)
+	}
+	e.dnsEntries[dnsName].namespaces[namespace] = struct{}{}
+	return e.dnsEntries[dnsName].dnsAddressSet, nil
+
+}
+
+func (e *EgressDNS) Delete(namespace string) bool {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	// go through all dnsNames for namespaces
+	for dnsName, dnsEntry := range e.dnsEntries {
+		// delete the dnsEntry
+		delete(dnsEntry.namespaces, namespace)
+		if len(dnsEntry.namespaces) == 0 {
+			// the dnsEntry appears in no other namespace so delete the address_set
+			err := dnsEntry.dnsAddressSet.Destroy()
+			if err != nil {
+				klog.Errorf("Error deleteing EgressFirewall AddressSet for dnsName: %s %v", dnsName, err)
+			}
+			// the dnsEntry is no longer needed because nothing referances it delete it
+			delete(e.dnsEntries, dnsName)
+		}
+	}
+	return len(e.dnsEntries) == 0
+}
+
+func (e *EgressDNS) Update(dns string) (bool, error) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
```

Comment:
@JacobTanenbaum I don't think we need locking here; e.dns.Update() does its own locking, and there's nothing we need to protect here about EgressDNS's Update right?

Reply:
+1

Reply:
done